### ISSUE TITLE
Wrongful execution of `ELSE` branch in `CASE` expression

### ISSUE THREAD

**Issue by CelineWuest** [neo4j/neo4j#13439]
I found a discrepancy when running two semantically equivalent queries against an empty database:

---
1:
```cypher
RETURN
  CASE
    WHEN null THEN COUNT { () }
  END AS x
```
Produces:
```
╒════╕
│x   │
╞════╡
│null│
└────┘
```

---
2:
```cypher
RETURN
  CASE
    WHEN (CASE WHEN null THEN COUNT { () } END) IS NULL THEN null
    WHEN (CASE WHEN null THEN COUNT { () } END) IS NOT NULL THEN (CASE WHEN null THEN COUNT { () } END)
    WHEN false THEN EXISTS { () }
    ELSE 1
  END AS x
```
Produces:
```
╒═══╕
│x  │
╞═══╡
│1  │
└───┘
```
---

I believe the queries mentioned above are semantically equivalent and thus their result should be the same.

The second query encases the case statement of the first and returns its result (`null` if it evaluates to `null`, otherwise just the case statement itself).  
Since one of the first two `WHEN` clauses in the second query must evaluate to true (since the encased expression is either `null` or not), the remaining branches shouldn't have an effect on the result, yet it seems like the database returned the value from the `ELSE` branch.

I encountered this issue when testing queries against the **Neo4j 5.18.0 enterprise version** in a Docker container running **alpine v.3**.
When testing the query against the community edition, the query results equal.

### Steps to reproduce
Run the aforementioned, semantically equivalent, queries against an empty database and observe that their results differ.

### Expected behavior
The queries should result in the same behavior

### Actual behavior
The query results differ

**Comment by LinneaAndersson** [comment]
Thanks for reporting this @DominicWuest, we will look into it and get back to you once we have more information.

**Comment by alexfoxgill** [comment]
thank you for the bug report, this bug will be fixed in version 5.21
